tensorflow/compiler/mlir/tosa/transforms/tosa_combine.cc

```diff
@@ -0,0 +1,183 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// Combine TOSA legalizations, eliminate redundancies.
+
+#include <climits>
+#include <cstddef>
+#include <cstdint>
+#include <iterator>
+#include <numeric>
+
+#include "mlir/Dialect/Quant/FakeQuantSupport.h"
+#include "mlir/Dialect/Quant/UniformSupport.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "mlir/Dialect/Tosa/Utils/QuantUtils.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/Module.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/StandardTypes.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "mlir/IR/Types.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "tensorflow/compiler/mlir/tensorflow/ir/tf_ops.h"
+#include "tensorflow/compiler/mlir/tosa/transforms/legalize_utils.h"
+
+#define PASS_NAME "tosa-combine"
+#define DEBUG_TYPE PASS_NAME
+
+namespace mlir {
+
+namespace tosa {
+
+namespace {
+
+class CombineTosa : public PassWrapper<CombineTosa, FunctionPass> {
+ public:
+  explicit CombineTosa() {}
+  void runOnFunction() override;
+};
+
+// This is meant to handle cases where we want to start from a newly legalized
+// TOSA op and further improve its original legalization.
+#define DECL_TOSACONVERT_OP(tosa_op)                                           \
+  struct ConvertTosa##tosa_op##Op : public RewritePattern {                    \
+    explicit ConvertTosa##tosa_op##Op(MLIRContext* context)                    \
+        : RewritePattern(tosa::tosa_op##Op::getOperationName(), 1, context) {} \
+    LogicalResult matchAndRewrite(Operation* op,                               \
+                                  PatternRewriter& rewriter) const override;   \
+  }
+DECL_TOSACONVERT_OP(Add);
+#undef DECL_TOSACONVERT_OP
+
+// Replaces the following pattern:
+//   %1 = tosa.conv2d (%ifm, %filter, %dummybias, ...)
+//   %3 = tosa.add(%1, %bias)
+//   with
+//   %1 = tosa.conv2d(%ifm, %filter, %bias, ...)
```

Comment:
Seems like this is a canonicalization that should go upstream? (I don't think it is TF specific)

Can this entire pass be moved upstream and used in TF?

Reply:
This pass is to address a TF specific issue where tf.Conv2D lack of bias as input. After discussing offline, we think this pass should fuse tf.Conv2D + tf.BiasAdd pattern, and legalize it to tosa.CONV2D directly. Will update this pass and rename this pass into --tosa-fuse-bias-tf in next patch

Reply:
Isn't the pattern exposed in the comment here something that can is a valid TOSA input and something you'd want to fuse in TOSA always?

Reply:
Yes that's true. But we didn't see this issue in any other dialect (e.g ONNX, TFLite) and that's why we eventually treat this more a TF specific issue.

It makes sense to fuse tf.Conv2d + tf.BiasAdd since tf.Conv2d doesn't have bias as input, while tosa.conv2d does. Fusing tosa.conv2d + tosa.add doesn't make much sense and what's worse that depends on the legalization of tf.Conv2d being tosa.conv2d and tf.BiasAdd being tosa.add, which adds another dependency.

Of course we can do tosa.conv2d + tosa.add like we did here. Certainly we could have that pattern to folding constant as an optimization pass, but that's not really the intention why we introduced this.

Hope this makes things clear!